File: 01/4-6.py
# !/usr/bin/env python3
import collections
import unittest

####### 1.4 #######
def simplifyWord (string):
    lst1 = string.lower()
    lst2 = list(lst1)
    if ' ' in lst2:
        lst2.remove(' ')
    return lst2

def myCounter(lst):
    dic = {}
    for i in lst:
        if i in dic: # has_key (only in Python2.X)
            dic[i] += 1
        else:
            dic[i] = 1
    return dic

def isPalindrome(lst):
    # counter = collections.Counter(lst) 
    counter = myCounter(lst)
    result = 0
    for i in counter.values():
        if i % 2 == 0:
            result = result
        else:
            result += 1
    if result <= 1:
        return True
    else:
        return False

####### 1.5 #######
def isOneStep(word1, word2):
    len1 = len(word1)
    len2 = len(word2)
    if len1 == len2:
        return same(word1, word2)
    elif len1 == len2 + 1:
        return longer(word1, word2)
    elif len1 + 1 == len2:
        return longer(word2, word1)
    else:
        return False

def same (word1, word2):
    if word1 == word2:
        return True
    else:
        lst1 = list(word1)
        lst2 = list(word2)
        cnt = 0
        cnt_rev = 0
        for i in range(len(word1)):
            if lst1[i] != lst2[i]:
                cnt = i
                break
        for j in range(len(word1)):
            if (lst1[::-1])[j] != (lst2[::-1])[j]:
                cnt_rev = j
                break
        if cnt + cnt_rev + 1 == len(word1):
            return True
        else:
            return False

def longer(word1, word2):
    for i in range(len(word1)):
        lst2 = list(word2)
        lst2.insert(i, '.')
        newword = ''.join(lst2)
        if same(word1, newword):
            return True
    return False 

####### 1.6 #######
def condense(word):
    lst = list(word)
    bucket = []
    result = ''
    for i in lst:
        if len(bucket) == 0:
            bucket.append([i, 1])
        elif bucket[-1][0] == i:
            bucket[-1][1] += 1
        else:
            bucket.append([i, 1])
    for i in bucket:
        result += i[0]
        result += str(i[1])
    if len(result) <= len(word):
        return result
    else:
        return word
            
####### test #######    
def test4(arg, expected):
    lst = simplifyWord(arg)
    answer = isPalindrome(lst)
    if answer == expected:
        print('PASS. ' + arg + ': ' + str(expected))
    else:
        print('FAIL. ' + arg + ' should be ' + str(expected) + ' but was ' + str(answer))

def test5(arg1, arg2, expected):
    answer = isOneStep(arg1, arg2)
    if answer == expected:
        print('PASS. ' + arg1 + ' and ' + arg2 + ': ' + str(expected))
    else:
        print('FAIL. ' + arg1 + ' and ' +  arg2 + ' should be ' + str(expected) + ' but was ' + str(answer))

def test6(arg, expected):
    answer = condense(arg)
    if answer == expected:
        print('PASS. ' + arg + ': ' + str(expected))
    else:
        print('FAIL. ' + arg + ' should be ' + expected + ' but was ' + answer)
        
def runTest():
    print('=== 1.4 started ====')
    test4('1', True)
    test4('a', True)
    test4(' ', True)
    test4('aa', True)
    test4('abc', False)
    test4('madam', True)
    test4('madam1', False)
    test4('ootee ', True)
    test4('Tact Coa', True)
    test4('TactCoaPAPA', True)
    print('=== 1.4 finished ===')
    print('=== 1.5 started ====')
    test5('same', 'same', True)
    test5('same', 'samee', True)
    test5('apple', ' apple', True)
    test5('apple', 'aple', True)
    test5('bake', 'pale', False)
    test5('hippo', 'elephant', False)
    print('=== 1.5 finished ===')
    print('=== 1.6 started ====')
    test6('a', 'a')
    test6('abc', 'abc')
    test6('aabccccc', 'a2b1c5')
    test6('ABC', 'ABC')
    test6('BBBBB', 'B5')
    test6('aabcd', 'aabcd') # a2b1c1d1 is longer
    test6('aabcccccaaa', 'a2b1c5a3')
    print('=== 1.6 finished ===')

if __name__ == '__main__':
    runTest()

File: 01/4-6.cpp
#include <iostream>
#include <string>
#include <vector>
#include <map>
#include <tuple>
using namespace std;

/////// 1.4 ////////
bool isPalindrome(string word){
    transform(word.begin(), word.end(), word.begin(), ::tolower);
    sort(word.begin(), word.end());
    int counter = 0;
    int result_counter = 0;
    for (int i = 0; i < word.length(); i++) {
        if (word[i] == ' '){
            continue;
        } else if (word[i] == word[i + 1]) {
            counter = (counter + 1) % 2;
        } else {
            counter = (counter + 1) % 2;
            if (counter == 1) {
                result_counter += 1;
                counter = 0;
            }
        }
    }
    if (result_counter <= 1){
        return true;
    } else {
        return false;
    }
}

/////// 1.5 ////////
bool isSame(string word1, string word2){
    if (word1 == word2) {
        return true;
    } else {
        int counter = 0;
        int counter_rev = 0;
        for (int i = 0; i < word1.length(); i++){
            if (word1[i] != word2[i]){
                counter = i;
                break;
            }
        }
        for (int j = 0; j < word1.length(); j++){
            string copied1 = word1;
            string copied2 = word2;
            reverse(copied1.begin(), copied1.end());
            reverse(copied2.begin(), copied2.end());
            if (copied1[j] != copied2[j]) {
                counter_rev = j;
                break;
            }
        }
        if (counter + counter_rev + 1 == word1.length()){
            return true;
        } else {
            return false;
        }
    }
}

bool isLonger(string word1, string word2){
    for (int i = 0; i < word1.length(); i++){
        string copied = word2; //local copy
        copied.insert(i, ".");
        if (isSame(word1, copied)){
            return true;
        }
    }
    return false;
}

bool isOneStep(string word1, string word2){
    int len1 = word1.length();
    int len2 = word2.length();
    if (len1 == len2){
        return isSame(word1, word2);
    } else if (len1 == len2 + 1){
        return isLonger(word1, word2);
    } else if (len1 + 1 == len2){
        return isLonger(word2, word1);
    } else {
        return false;
    }
}

/////// 1.6 ///////
// iterator
string condense(string word){
    string result = "";
    for(string::iterator itr = word.begin(); itr != word.end(); ){
        int counter = 0;
        string::iterator tmp = itr;
        while (*itr == *tmp){
            counter++;
            tmp++;
        }
        result += *itr;
        result += to_string(counter);
        itr = tmp;
    }
    if (word.length() <= result.length()){
        return word;
    } else {
        return result;
    }
}


/////// tests ///////
void test4(string arg, bool expected){
    bool answer = isPalindrome(arg);
    cout << boolalpha; // to show "true" instead of "1"
    if (answer == expected) {
        cout << "PASS. " << arg << ": " << expected << "\n";
    } else {
        cout << "FAIL. " << arg << " should be " << expected << " but was " << answer << "\n";
    }
}

void test5(string arg1, string arg2, bool expected){
    bool answer = isOneStep(arg1, arg2);
    if (answer == expected) {
        cout << boolalpha; // to show "true" instead of "1"
        cout << "PASS. " << arg1 << " and " << arg2 << ": " << expected << "\n";
    } else {
        cout << boolalpha;
        cout << "FAIL. " << arg1 << " and " << arg2 << " should be " << expected << " but was " << answer << "\n";
    }
}

void test6(string arg, string expected){
    string answer = condense(arg);
    if (answer == expected){
        cout << "PASS. " << arg << ": " << expected << "\n";
    } else {
        cout << "FAIL. " << arg << " should be " << expected;
        cout << " but was " << answer << "\n";
    }
}

int main(){
    std::cout << "=== 1.4 started ====\n";
    test4("1", true);
    test4("a", true);
    test4(" ", true);
    test4("aa", true);
    test4("abc", false);
    test4("madam", true);
    test4("madam1", false);
    test4("ootee ", true);
    test4("Tact Coa", true);
    test4("TactCoaPAPA", true);
    std::cout << "=== 1.4 finished ===\n";
    std::cout << "=== 1.5 started ====\n";
    test5("same", "same", true);
    test5("same", "samee", true);
    test5("apple", " apple", true);
    test5("apple", "aple", true);
    test5("bake", "pale", false);
    test5("hippo", "elephant", false);
    std::cout << "=== 1.5 finished ===\n";
    std::cout << "=== 1.6 started ====\n";
    test6("a", "a");
    test6("abc", "abc");
    test6("aabccccc", "a2b1c5");
    test6("ABC", "ABC");
    test6("BBBBB", "B5");
    test6("aabcd", "aabcd"); // a2b1c1d1 is longer
    test6("aabcccccaaa", "a2b1c5a3");
    std::cout << "=== 1.6 finished ===\n";
    return 0;
};

